File: app/src/main/java/ru/shprot/sudokumobdevkz/model/game/generator/Solver/DancingLinks.java
package ru.shprot.sudokumobdevkz.model.game.generator.Solver;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import ru.shprot.sudokumobdevkz.model.game.generator.Generator;


public class DancingLinks {

    private final ColumnNode header;
    private List<DancingNode> answer;
    public static int solutionsCount;

    private void search(int k) {
        if (solutionsCount == 2) return;
        if (header.R == header) {
            solutionsCount++;
            Generator.isUnicSolution = solutionsCount == 1;
        } else {
            ColumnNode c = selectColumnNodeHeuristic();
            c.cover();

            for (DancingNode r = c.D; r != c; r = r.D) {
                answer.add(r);

                for (DancingNode j = r.R; j != r; j = j.R) {
                    j.C.cover();
                }

                search(k + 1);

                r = answer.remove(answer.size() - 1);
                c = r.C;

                for (DancingNode j = r.L; j != r; j = j.L) {
                    j.C.uncover();
                }
            }
            c.uncover();
        }
    }

    private ColumnNode selectColumnNodeHeuristic() {
        int min = Integer.MAX_VALUE;
        ColumnNode ret = null;
        for (ColumnNode c = (ColumnNode) header.R; c != header; c = (ColumnNode) c.R) {
            if (c.size < min) {
                min = c.size;
                ret = c;
            }
        }
        return ret;
    }

    private ColumnNode makeDLXBoard(boolean[][] grid) {
        final int COLS = grid[0].length;

        ColumnNode headerNode = new ColumnNode("header");
        List<ColumnNode> columnNodes = new ArrayList<>();

        for (int i = 0; i < COLS; i++) {
            ColumnNode n = new ColumnNode(Integer.toString(i));
            columnNodes.add(n);
            headerNode = (ColumnNode) headerNode.hookRight(n);
        }
        headerNode = headerNode.R.C;

        for (boolean[] aGrid : grid) {
            DancingNode prev = null;
            for (int j = 0; j < COLS; j++) {
                if (aGrid[j]) {
                    ColumnNode col = columnNodes.get(j);
                    DancingNode newNode = new DancingNode(col);
                    if (prev == null)
                        prev = newNode;
                    col.U.hookDown(newNode);
                    prev = prev.hookRight(newNode);
                    col.size++;
                }
            }
        }

        headerNode.size = COLS;

        return headerNode;
    }

    DancingLinks(boolean[][] cover) {
        header = makeDLXBoard(cover);
        solutionsCount = 0;
    }

    public void runSolver() {
        answer = new LinkedList<>();
        search(0);
    }

}

File: app/src/main/java/ru/shprot/sudokumobdevkz/model/generator/solver/DancingLinks.kt
package ru.shprot.sudokumobdevkz.model.generator.solver

import java.util.LinkedList

internal class DancingLinks(cover: BooleanArray2D) {

    private val header: ColumnNode = makeDLXBoard(cover)
    private var answer: MutableList<DancingNode> = LinkedList()

    var solutionsCount: Int = 0
        private set

    fun runSolver() {
        answer = LinkedList()
        solutionsCount = 0
        search(0)
    }

    private fun search(k: Int) {
        if (solutionsCount >= 2) return
        if (header.R == header) {
            solutionsCount++
            return
        }

        val c = selectColumnNodeHeuristic() ?: return
        c.cover()

        var r = c.D
        while (r != c) {
            answer.add(r)

            var j = r.R
            while (j != r) {
                (j.C as ColumnNode).cover()
                j = j.R
            }

            search(k + 1)

            r = answer.removeAt(answer.size - 1)
            val col = r.C as ColumnNode

            j = r.L
            while (j != r) {
                (j.C as ColumnNode).uncover()
                j = j.L
            }

            r = r.D
        }

        c.uncover()
    }

    private fun selectColumnNodeHeuristic(): ColumnNode? {
        var min = Int.MAX_VALUE
        var ret: ColumnNode? = null
        var c = header.R.C as? ColumnNode
        while (c != null && c != header) {
            if (c.size < min) {
                min = c.size
                ret = c
            }
            c = c.R.C as? ColumnNode
        }
        return ret
    }

    private fun makeDLXBoard(grid: BooleanArray2D): ColumnNode {
        val cols = grid.cols

        var headerNode = ColumnNode("header")
        val columnNodes = mutableListOf<ColumnNode>()

        for (i in 0 until cols) {
            val n = ColumnNode(i.toString())
            columnNodes.add(n)
            headerNode = headerNode.hookRight(n).C as ColumnNode
        }
        headerNode = headerNode.R.C as ColumnNode

        for (row in 0 until grid.rows) {
            var prev: DancingNode? = null
            for (j in 0 until cols) {
                if (grid[row, j]) {
                    val col = columnNodes[j]
                    val newNode = DancingNode(col)
                    if (prev == null) prev = newNode
                    col.U.hookDown(newNode)
                    prev = prev.hookRight(newNode)
                    col.size++
                }
            }
        }

        headerNode.size = cols
        return headerNode
    }
}
